Add migration guide (#432)
cloudant/nodejs-cloudant: MIGRATION.md

# Migrating to the `cloudant-node-sdk` library

> This document is to assist in migrating from the `nodejs-cloudant` (package: `@cloudant/cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`) that is JavaScript and TypeScript compliant.

Reviewer: Out of curiosity, why did you change compatible with compliant? Compatible seems to be the better word for the relationship between `cloudant-node-sdk` and JS/TS.

Author: Okay, then I revert it back :)

Author: Changed in 40dd63a

## Initializing the client connection
There are several ways to create a client connection in `cloudant-node-sdk`:
1. [Environment variables](https://github.com/IBM/cloudant-node-sdk#authentication-with-environment-variables)
2. [External configuration file](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration)
3. [Programmatically](https://github.com/IBM/cloudant-node-sdk#programmatic-authentication)

[See the README](https://github.com/IBM/cloudant-node-sdk#code-examples) for code examples on using environment variables.

## Other differences

> 1. The way of storing credentials in a file has changed, the usage of `dotenv` package is not suggested. See the [external file configuration section](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration) in our API docs about the new handling.

Author: Yes, this looks good now. Changed: 583c25f

> 1. Fetching the database object first before performing operations on it is not required. For example, in the case of updating a document, this would mean calling first `getDocument` to fetch and after `putDocument` to update.

Reviewer: `before performing operations on it` sounds a bit awkward.  What about:
```suggestion
1. Fetching the database object first before performing additional operations is not required. For example, in the case of updating a document you would first call `getDocument` to fetch and then `putDocument` to update.
```

Author: Okay. Changed 583c25f

|`disableReplication()`|[deleteReplicationDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#deletereplicationdocument)|
|`session()`|[getSessionInformation](https://cloud.ibm.com/apidocs/cloudant?code=node#getsessioninformation)|
|`uuids()`|[getUuids](https://cloud.ibm.com/apidocs/cloudant?code=node#getuuids)|
|`db.destroy()`|[deleteDatabase](https://cloud.ibm.com/apidocs/cloudant?code=node#deletedatabase)|
|`db.info()`|[getDatabaseInformation](https://cloud.ibm.com/apidocs/cloudant?code=node#getdatabaseinformation)|
|`db.insert()`|[postDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#postdocument)|
|`db.create(db_name)`|[putDatabase](https://cloud.ibm.com/apidocs/cloudant?code=node#putdatabase)|
|`db.fetch()/db.list()/db.listAsStream()`|[postAllDocs](https://cloud.ibm.com/apidocs/cloudant?code=node#postalldocs)|
|`db.bulk()`|[postBulkDocs](https://cloud.ibm.com/apidocs/cloudant?code=node#postbulkdocs)|
|`db.bulk_get()`|[postBulkGet](https://cloud.ibm.com/apidocs/cloudant?code=node#postbulkget)|
|`db.changes()`|[postChanges](https://cloud.ibm.com/apidocs/cloudant?code=node#postchanges)|

> |`db.destroy()`|[deleteDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#deletedesigndocument)|

Reviewer: I think we'll want to include the _design path for these operations:
```suggestion
|`db.destroy() with _design path`|[deleteDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#deletedesigndocument)|
```

Author: Changed in 583c25f

> |`db.get()`|[getDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#getdesigndocument)|

Reviewer: ```suggestion
|`db.get() with _design path`|[getDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#getdesigndocument)|
```

Author: Changed in 583c25f

> |`db.insert()`|[putDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#putdesigndocument)|

Reviewer: ```suggestion
|`db.insert() with _design path`|[putDesignDocument](https://cloud.ibm.com/apidocs/cloudant?code=node#putdesigndocument)|
```

Author: Changed in 583c25f

# Migrating to the `cloudant-node-sdk` library

> This document is to assist in migrating from the `nodejs-cloudant` (module: `cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`). Compatible with Javascript and TypeScript.

Reviewer 2: Nit:
```suggestion
This document is to assist in migrating from the `nodejs-cloudant` (module: `cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`) that is supported in Javascript and TypeScript.
```

Author: This way `supported` word is used 2 times in a sentence. On the Hungarian Literature lesson taught me this is an invalid composition, better to use synonyms.
Okay, this is not a novel, but still. 
WDYT about this:
```suggestion
This document is to assist in migrating from the `nodejs-cloudant` (module: `cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`) that is applicable with Javascript and TypeScript.
```

Reviewer 2: Good call. I'm not an English teacher though :)

Author: Yeah, I checked this word, it is rather used in legal context. So switched back to `compatible` currently in c706430

Author: Today I found an other word: `compliant` WDYT?

Author: Check it here: 583c25f

Author: Reverted this back: 40dd63a

> This document is to assist in migrating from the `nodejs-cloudant` (module: `cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`). Compatible with Javascript and TypeScript.

Reviewer: Let's use package for both libraries, and the legacy package is named `@cloudant/cloudant`
```suggestion
This document is to assist in migrating from the `nodejs-cloudant` (package: `@cloudant/cloudant`) to the newly supported [`cloudant-node-sdk`](https://github.com/IBM/cloudant-node-sdk) (package: `@ibm-cloud/cloudant`). Compatible with Javascript and TypeScript.
```

Author: Changed in c706430

## Initializing the client connection
There are several ways to create a client connection in `cloudant-node-sdk`:
1. [Environment variables](https://github.com/IBM/cloudant-node-sdk#authentication-with-environment-variables)
2. [External configuration file](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration)
3. [Programmatically](https://github.com/IBM/cloudant-node-sdk#programmatic-authentication)

[See the README](https://github.com/IBM/cloudant-node-sdk#code-examples) for code examples on using environment variables.

## Other differences

> 1. For storing credentials in a file, `dotenv` package not needed anymore. Check the new operation [here](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration).

Reviewer: How about:
```suggestion
1. For storing credentials in a file, the `dotenv` package is not supported. See the [external file configuration section](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration) in our API docs for handling this feature in our new library.
```

Author: For me `the `dotenv` package is not supported.` says that we deal with issues related to `dotenv` but this is not the case. This is just a suggested package in this situations. So I am reluctant about the first part. 

WDYT about this:
```suggestion
1. For storing credentials in a file has changed, the `dotenv` package is not needed anymore. See the [external file configuration section](https://github.com/IBM/cloudant-node-sdk#authentication-with-external-configuration) in our API docs for the new handling.
```

Author: Changed in c706430

> 1. No need to fetch database object in advance to perform operations on Documents. For example, in the case of the document update, this would mean calling first `getDocument` to fetch and after `putDocument` to update.

Reviewer: Does this apply to just documents?
```suggestion
1. Fetching the database object first before performing operations on documents is not required. For example, in the case of updating a document, this would mean calling first `getDocument` to fetch and after `putDocument` to update.
```

Author: Not really, there are a bunch of other things like: replicate, db info, view, find.
I also removed `on documents` from the first sentence.
```suggestion
1. Fetching the database object first before performing operations on it is not required. For example, in the case of updating a document, this would mean calling first `getDocument` to fetch and after `putDocument` to update.
```

Author: Changed in c706430

> 1. There are no plugin handling anymore.

Reviewer: We might want to say that plugin handling isn't supported but several of the plugins exist in the new library.  We might also want to give an example of a plugin (e.g. IAM) that is supported e.g.
```suggestion
1. Plugins are not supported, but several of the plugin features exist in the new library e.g. IAM.
```

Author: Changed in c706430

> 1. Authentication mode can be set [this way](https://github.com/IBM/cloudant-node-sdk/tree/master#authentication).

Reviewer: I think we can remove this since we already have a section about for authentication.

Author: Removed in c706430

> 1. Retries currently not supported.

Reviewer: Let's be a bit more specific about the retry feature:
```suggestion
1. The plugin for retrying failed requests is not supported.
```

Author: Changed in c706430
